File: wire/src/source.rs
use crate::agent::wire_id;
use crate::error::InvalidError;
use crate::validate::Validate;
use serde::{Deserialize, Serialize};
use std::collections::BTreeSet;

pub const MAX_SOURCE_PARTITIONS: usize = 65_536;

wire_id!(
    /// Durable identity of one physical Iggy cluster across replica replacement and restore.
    PhysicalClusterIncarnation
);

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub struct SourceIncarnation {
    pub cluster: PhysicalClusterIncarnation,
    pub stream_id: u32,
    pub topic_id: u32,
    pub partition_id: u32,
    pub partition_created_revision: u64,
}

impl Validate for SourceIncarnation {
    fn validate(&self) -> Result<(), InvalidError> {
        if self.cluster.as_u128() == 0 {
            return Err(InvalidError::new(
                "physical cluster incarnation must be nonzero",
            ));
        }
        if self.partition_created_revision == 0 {
            return Err(InvalidError::new(
                "partition created revision must be nonzero",
            ));
        }
        Ok(())
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct SourceScope {
    pub stream: String,
    pub topic: String,
}

impl SourceScope {
    pub fn new(stream: impl Into<String>, topic: impl Into<String>) -> Self {
        Self {
            stream: stream.into(),
            topic: topic.into(),
        }
    }
}

impl Validate for SourceScope {
    fn validate(&self) -> Result<(), InvalidError> {
        validate_source_name("stream", &self.stream)?;
        validate_source_name("topic", &self.topic)
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct SourcePartitionCut {
    pub incarnation: SourceIncarnation,
    pub retained_start: u64,
    pub end_exclusive: u64,
}

impl Validate for SourcePartitionCut {
    fn validate(&self) -> Result<(), InvalidError> {
        self.incarnation.validate()?;
        if self.retained_start > self.end_exclusive {
            return Err(InvalidError::new(format!(
                "retained start {} is above end-exclusive head {}",
                self.retained_start, self.end_exclusive
            )));
        }
        Ok(())
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct SourceCut {
    pub metadata_revision: u64,
    pub partitions: Vec<SourcePartitionCut>,
}

impl Validate for SourceCut {
    fn validate(&self) -> Result<(), InvalidError> {
        if self.metadata_revision == 0 {
            return Err(InvalidError::new(
                "source metadata revision must be nonzero",
            ));
        }
        if self.partitions.is_empty() || self.partitions.len() > MAX_SOURCE_PARTITIONS {
            return Err(InvalidError::new(format!(
                "source cut must contain 1..={MAX_SOURCE_PARTITIONS} partitions"
            )));
        }
        let mut partitions = BTreeSet::new();
        let mut previous_partition = None;
        let mut namespace = None;
        for partition in &self.partitions {
            partition.validate()?;
            let current = (
                partition.incarnation.cluster,
                partition.incarnation.stream_id,
                partition.incarnation.topic_id,
            );
            if namespace.is_some_and(|expected| expected != current) {
                return Err(InvalidError::new(
                    "source cut partitions must share one cluster, stream, and topic",
                ));
            }
            namespace = Some(current);
            if !partitions.insert(partition.incarnation.partition_id) {
                return Err(InvalidError::new(format!(
                    "source cut repeats partition {}",
                    partition.incarnation.partition_id
                )));
            }
            if previous_partition
                .is_some_and(|previous| previous >= partition.incarnation.partition_id)
            {
                return Err(InvalidError::new(
                    "source cut partitions must be ordered by ascending partition id",
                ));
            }
            previous_partition = Some(partition.incarnation.partition_id);
        }
        Ok(())
    }
}

fn validate_source_name(label: &str, value: &str) -> Result<(), InvalidError> {
    if value.is_empty() {
        return Err(InvalidError::new(format!("{label} must not be empty")));
    }
    if value.len() > 255 {
        return Err(InvalidError::new(format!(
            "{label} is {} bytes, exceeds cap 255",
            value.len()
        )));
    }
    if value.chars().any(char::is_control) {
        return Err(InvalidError::new(format!(
            "{label} contains a control character"
        )));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn incarnation(partition_id: u32, created_revision: u64) -> SourceIncarnation {
        SourceIncarnation {
            cluster: PhysicalClusterIncarnation::from_u128(1),
            stream_id: 10,
            topic_id: 20,
            partition_id,
            partition_created_revision: created_revision,
        }
    }

    #[test]
    fn given_a_recreated_partition_when_incarnations_are_compared_then_should_differ() {
        assert_ne!(incarnation(1, 7), incarnation(1, 8));
        incarnation(1, 7).validate().expect("incarnation is valid");
    }

    #[test]
    fn given_duplicate_partitions_or_inverted_heads_when_a_source_cut_is_validated_then_should_reject()
     {
        let duplicate = SourceCut {
            metadata_revision: 9,
            partitions: vec![
                SourcePartitionCut {
                    incarnation: incarnation(1, 7),
                    retained_start: 0,
                    end_exclusive: 10,
                },
                SourcePartitionCut {
                    incarnation: incarnation(1, 7),
                    retained_start: 2,
                    end_exclusive: 10,
                },
            ],
        };
        assert!(duplicate.validate().is_err());

        let inverted = SourcePartitionCut {
            incarnation: incarnation(1, 7),
            retained_start: 11,
            end_exclusive: 10,
        };
        assert!(inverted.validate().is_err());
    }
}

File: foreign/typescript/src/wire/source.ts
import { InvalidError } from "../client/errors.js"
import { type CborMap, expectMap, field } from "./cbor.js"
import { PhysicalClusterIncarnation } from "./ids.js"

export const MAX_SOURCE_PARTITIONS = 65_536
export interface SourceIncarnation {
  readonly cluster: PhysicalClusterIncarnation
  readonly streamId: number
  readonly topicId: number
  readonly partitionId: number
  readonly partitionCreatedRevision: bigint
}
export interface SourceScope {
  readonly stream: string
  readonly topic: string
}
export interface SourcePartitionCut {
  readonly incarnation: SourceIncarnation
  readonly retainedStart: bigint
  readonly endExclusive: bigint
}
export interface SourceCut {
  readonly partitions: readonly SourcePartitionCut[]
}

export function encodeSourceIncarnation(value: SourceIncarnation): Map<string, unknown> {
  return new Map<string, unknown>([
    ["cluster", value.cluster.toBytes()],
    ["stream_id", value.streamId],
    ["topic_id", value.topicId],
    ["partition_id", value.partitionId],
    ["partition_created_revision", value.partitionCreatedRevision]
  ])
}
export function decodeSourceIncarnation(map: CborMap, context: string): SourceIncarnation {
  return {
    cluster: PhysicalClusterIncarnation.fromBytes(field.requiredBytes(map, "cluster", context)),
    streamId: field.requiredU32(map, "stream_id", context),
    topicId: field.requiredU32(map, "topic_id", context),
    partitionId: field.requiredU32(map, "partition_id", context),
    partitionCreatedRevision: field.requiredU64(map, "partition_created_revision", context)
  }
}
export function encodeSourceScope(value: SourceScope): Map<string, unknown> {
  return new Map([
    ["stream", value.stream],
    ["topic", value.topic]
  ])
}
export function decodeSourceScope(map: CborMap, context: string): SourceScope {
  return {
    stream: field.requiredString(map, "stream", context),
    topic: field.requiredString(map, "topic", context)
  }
}
export function encodeSourcePartitionCut(value: SourcePartitionCut): Map<string, unknown> {
  return new Map<string, unknown>([
    ["incarnation", encodeSourceIncarnation(value.incarnation)],
    ["retained_start", value.retainedStart],
    ["end_exclusive", value.endExclusive]
  ])
}
export function decodeSourcePartitionCut(map: CborMap, context: string): SourcePartitionCut {
  return {
    incarnation: decodeSourceIncarnation(
      field.requiredMap(map, "incarnation", context),
      `${context}.incarnation`
    ),
    retainedStart: field.requiredU64(map, "retained_start", context),
    endExclusive: field.requiredU64(map, "end_exclusive", context)
  }
}
export function encodeSourceCut(value: SourceCut): Map<string, unknown> {
  return new Map<string, unknown>([["partitions", value.partitions.map(encodeSourcePartitionCut)]])
}
export function decodeSourceCut(map: CborMap, context: string): SourceCut {
  return {
    partitions: field.requiredArray(map, "partitions", context, (item, index) =>
      decodeSourcePartitionCut(
        expectMap(item, `${context}.partitions[${String(index)}]`),
        `${context}.partitions[${String(index)}]`
      )
    )
  }
}
export function validateSourceIncarnation(value: SourceIncarnation): void {
  if (value.cluster.asU128() === 0n || value.partitionCreatedRevision === 0n) {
    throw new InvalidError("source incarnation identity or creation revision is invalid")
  }
}
export function validateSourceScope(value: SourceScope): void {
  validateSourceName("stream", value.stream)
  validateSourceName("topic", value.topic)
}
export function validateSourceCut(value: SourceCut): void {
  if (value.partitions.length < 1 || value.partitions.length > MAX_SOURCE_PARTITIONS)
    throw new InvalidError("source cut revision or partition count is invalid")
  let previous = -1
  let namespace: string | undefined
  for (const partition of value.partitions) {
    validateSourceIncarnation(partition.incarnation)
    if (
      partition.incarnation.cluster.asU128() === 0n ||
      partition.incarnation.partitionCreatedRevision === 0n ||
      partition.retainedStart > partition.endExclusive ||
      partition.incarnation.partitionId <= previous
    )
      throw new InvalidError("source cut partition is invalid or not canonically ordered")
    const current = `${partition.incarnation.cluster.asU128().toString()}:${String(partition.incarnation.streamId)}:${String(partition.incarnation.topicId)}`
    if (namespace !== undefined && namespace !== current)
      throw new InvalidError("source cut partitions must share one source namespace")
    namespace = current
    previous = partition.incarnation.partitionId
  }
}

function validateSourceName(label: string, value: string): void {
  const bytes = new TextEncoder().encode(value).length
  if (bytes < 1 || bytes > 255 || hasControlCharacter(value)) {
    throw new InvalidError(`${label} is empty, oversized, or contains a control character`)
  }
}

function hasControlCharacter(value: string): boolean {
  for (const character of value) {
    const code = character.codePointAt(0) ?? 0
    if (code < 32 || code === 127) return true
  }
  return false
}
